internal/core/src/storage/PayloadWriter.cpp

```diff
@@ -0,0 +1,106 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/PayloadWriter.h"
+#include "exceptions/EasyAssert.h"
+#include "common/FieldMeta.h"
+#include "storage/Util.h"
+
+namespace storage {
+
+// create payload writer for numeric data type
+// lazy init for float/binary vector
+// Compatible with codec logic implemented by go
+PayloadWriter::PayloadWriter(const DataType column_type) : column_type_(column_type) {
+    builder_ = CreateArrowBuilder(column_type);
+    schema_ = CreateArrowSchema(column_type);
+}
+
+// create payload writer for vector data type
+PayloadWriter::PayloadWriter(const DataType column_type, int dim) : column_type_(column_type) {
+    init_dimension(dim);
+}
+
+void
+PayloadWriter::init_dimension(int dim) {
+    if (dimension_.has_value()) {
+        AssertInfo(dimension_ == dim, "init dimension with diff values repeatedly");
+        return;
+    }
+
+    dimension_ = dim;
+    builder_ = CreateArrowBuilder(column_type_, dim);
+    schema_ = CreateArrowSchema(column_type_, dim);
+}
+
+void
+PayloadWriter::add_one_string_payload(const char* str, int str_size) {
+    AssertInfo(output_ == nullptr, "payload writer has been finished");
+    AssertInfo(milvus::datatype_is_string(column_type_), "mismatch data type");
+    AddOneStringToArrowBuilder(builder_, str, str_size);
+    rows_.fetch_add(1);
+}
+
+void
+PayloadWriter::add_payload(const Payload& raw_data) {
+    AssertInfo(output_ == nullptr, "payload writer has been finished");
+    AssertInfo(column_type_ == raw_data.data_type, "mismatch data type");
+
+    // lazy init for float/binary vector
+    // Compatible with codec logic implemented by go
+    if (milvus::datatype_is_vector(column_type_) && !dimension_.has_value()) {
+        AssertInfo(raw_data.dimension.has_value(), "empty dimension");
+        init_dimension(raw_data.dimension.value());
```

Comment:
should we add some checker to find errors early, like checking the size of raw data is right